eeps/eep-0054.md

```diff
@@ -0,0 +1,361 @@
+    Author: Björn Gustavsson <bjorn(at)erlang(dot)org>
+    Status: Draft
+    Type: Standards Track
+    Created: 14-Sep-2020
+    Erlang-Version: 24
+    Post-History:
+****
+EEP 54: Provide more information about errors
+----
+
+
+
+Abstract
+========
+
+This EEP proposes a mechanism for reporting more information about
+what went wrong when a BIF raises an exception.  The same mechanism
+can be used by libraries or applications to provide more detailed
+error messages.
+
+
+
+
+Specification
+=============
+
+In OTP 23 and earlier, the shell prints a terse message when a call
+to a built-in function (BIF) fails:
+
+    1> element(a,b).
+    ** exception error: bad argument
+         in function  element/2
+            called as element(a,b)
+
+The `bad argument` message informs us that one or more of the
+arguments to the call were incorrect in some way (in this example,
+both arguments have wrong types).
+
+We propose a mechanism that enables the shell to print more helpful
+error messages.  Here is how the message would be printed with the
+reference implementation of this EEP:
+
+    1> element(a, b).
+    ** exception error: bad argument
+         in function  element/2
+            called as element(a,b)
+            *** argument 1: not an integer
+            *** argument 2: not a tuple
+
+Note that the exact formatting and phrasing of the messages is an implementation
+detail outside the scope of this EEP.  What will be specified here are the APIs
+and conventions that make these messages possible.
+
+
+### Proposals in this EEP
+
+* An extension of the format of the call-stack back trace (**stacktrace**) format
+  to indicate that there exists extended error information for that call, and
+  a convention for how extended error information can be provided.
+
+* A new `erlang:error/3` BIF to allow libraries and applications to
+  raise an exception with extended error information.
+
+* New functions `erl_error:format_exception/3` and
+  `erl_error:format_exception/4` to allow libraries and applications to
+  format stacktraces in the same style as the shell.
+
+
+### Extending the stacktrace
+
+The stack back-trace (**stacktrace**) is currently a list of tuples.
+For the purpose of this EEP we are only interested in the first entry
+in the stacktrace.  It looks like
+`{Module,Function,Arguments,ExtraInfo}`, where `ExtraInfo` is a list of
+two-tuples.
+
+As an indication that extended error info is available, we propose
+adding an `{error_info,ErrorInfo}` tuple to `ExtraInfo` in the
+first element in the stacktrace.  If that tuple is present when an
+exception is caught, it indicates that `Module` has a
+`format_error/3` function that can be called to provide additional
+information about the error.  The arguments for `format_error/3` are
+`Function`, `Arguments`, and `ErrorInfo` from the `error_info` tuple.
+
+Thus, if the first entry in the stacktrace is:
+
+    {erlang,element,[1,no_tuple],[{error_info,ErrorInfo}]}
```

Comment:
```suggestion
    {erlang,element,[1,no_tuple],[{error_info,#{ info => ErrorInfo }}]}
```

In order to be able to extend the meta-data associated with `error_info`, would it make sense to also make it a map?

The extension I have in mind is allowing `#{ module => M, function => F, info => ErrorInfo }` in the future (or now).

Reply:
Yes, that make sense.

Reply:
I have now updated the reference implementation and the EEP to have a map as the second element in the `error_info` tuple. (I chose to use the key `reason` instead of `info`.)